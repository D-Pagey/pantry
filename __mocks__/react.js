/* eslint-disable global-require */
const reactActual = jest.requireActual('react');

module.exports = {
    ...reactActual,
    // @TODO: remove when warnings are fixed: https://github.com/kentcdodds/react-testing-library/issues/281#issuecomment-465626540
    useState: (initialState) => {
        const [state, setState] = reactActual.useState(initialState);
        return [
            state,
            (value) => {
                require('react-dom/test-utils').act(() => {
                    setState(value);
                });
            }
        ];
    }
};
