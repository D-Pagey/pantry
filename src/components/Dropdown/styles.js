import styled from 'styled-components';

export const Wrapper = styled.div`
    display: flex;
    flex-direction: column;
    margin: 0 0 1rem;
`;

export const Label = styled.label`
    margin: 0 0 0.5rem;
`;
