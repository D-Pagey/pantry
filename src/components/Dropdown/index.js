import React from 'react';
import { arrayOf, func, shape, string } from 'prop-types';
import Select from 'react-select';
import * as S from './styles';

const Dropdown = ({ error, label, options, selected, setSelected }) => {
    const handleChange = (value) => setSelected(value);

    return (
        <S.Wrapper>
            {label && <S.Label>{label}</S.Label>}

            <Select
                options={options}
                onChange={handleChange}
                value={selected}
                isClearable
                isSearchable
            />

            {error && <S.Error>{error}</S.Error>}
        </S.Wrapper>
    );
};

Dropdown.propTypes = {
    error: string,
    label: string,
    options: arrayOf(
        shape({
            label: string.isRequired,
            value: string.isRequired
        }).isRequired
    ).isRequired,
    selected: shape({
        label: string,
        value: string
    }),
    setSelected: func.isRequired
};

Dropdown.defaultProps = {
    error: '',
    label: '',
    selected: undefined
};

export default Dropdown;
