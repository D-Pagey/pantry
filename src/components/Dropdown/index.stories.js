import React, { useEffect, useState } from 'react';
import { storiesOf } from '@storybook/react';
import Dropdown from '.';

const options = [
    {
        label: 'Meat',
        value: 'meat'
    },
    {
        label: 'Vegetables',
        value: 'vegetables'
    },
    {
        label: 'Fish',
        value: 'fish'
    }
];

// eslint-disable-next-line react/prop-types
const FakeForm = ({ error, label, preSelected }) => {
    const [selected, setSelected] = useState();

    useEffect(() => {
        if (preSelected) setSelected(preSelected);
    }, [preSelected]);

    return (
        <Dropdown
            error={error}
            options={options}
            label={label}
            selected={selected}
            setSelected={setSelected}
        />
    );
};

storiesOf('Dropdown', module)
    .add('with label', () => <FakeForm label="What category of food?" />)
    .add('without label', () => <FakeForm />)
    .add('preselected', () => <FakeForm preSelected={options[1]} />)
    .add('with error', () => <FakeForm preSelected={options[1]} error="Required" />);
