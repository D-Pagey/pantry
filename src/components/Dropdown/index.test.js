/* eslint-disable react/prop-types */
import React from 'react';
import { wait, fireEvent } from '@testing-library/react';
import Dropdown from '.';

jest.mock('react-select', () => ({ options, value, onChange }) => {
    const handleChange = (event) => {
        const option = options.find((item) => item.value === event.currentTarget.value);
        onChange(option);
    };

    return (
        <select data-testid="select" value={value} onChange={handleChange}>
            {options.map((item) => (
                <option key={item.value} value={item.value}>
                    {item.label}
                </option>
            ))}
        </select>
    );
});

const props = {
    options: [
        {
            label: 'Meat',
            value: 'meat'
        },
        {
            label: 'Vegetables',
            value: 'vegetables'
        },
        {
            label: 'Fish',
            value: 'fish'
        }
    ],
    setSelected: () => {}
};

describe('Dropdown component', () => {
    it('should render', () => {
        const { container } = render(<Dropdown {...props} />);
        expect(container.firstChild).toMatchSnapshot();
    });

    it('should render a label', () => {
        const label = 'What food category is it?';
        const { getByText } = render(<Dropdown {...props} label={label} />);
        getByText(label);
    });

    it('should handle change', async () => {
        const setSelected = jest.fn();
        const { getByTestId } = render(<Dropdown {...props} setSelected={setSelected} />);

        fireEvent.change(getByTestId('select'), { target: { value: 'vegetables' } });

        await wait(() =>
            expect(setSelected).toHaveBeenCalledWith({ label: 'Vegetables', value: 'vegetables' })
        );
    });
});
