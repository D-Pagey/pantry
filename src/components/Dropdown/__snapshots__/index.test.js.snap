// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Dropdown component should render 1`] = `
.c0 {
  display: -webkit-box;
  display: -webkit-flex;
  display: -ms-flexbox;
  display: flex;
  -webkit-flex-direction: column;
  -ms-flex-direction: column;
  flex-direction: column;
  margin: 0 0 1rem;
}

<div
  class="c0"
>
  
  <select
    data-testid="select"
  >
    <option
      value="meat"
    >
      Meat
    </option>
    <option
      value="vegetables"
    >
      Vegetables
    </option>
    <option
      value="fish"
    >
      Fish
    </option>
  </select>
  
</div>
`;
